runtime/src/issuance.rs

```diff
@@ -0,0 +1,39 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// A trait for types that can provide the amount of issuance to award to the block
+/// author for the given block number.
+pub trait Issuance<BlockNumber, Balance> {
+	fn issuance(block: BlockNumber) -> Balance;
+}
+
+// A minimal implementation for when you don't actually want any issuance
+impl Issuance<u32, u128> for () {
+	fn issuance(_block: u32) -> u128 {
+		0
+	}
+}
+
+/// A type that provides block issuance according to bitcoin's rules
+/// Initial issuance is 50 / block
+/// Issuance is cut in half every 210,000 blocks
+/// cribbed from github.com/Bitcoin-ABC/bitcoin-abc/blob/9c7b12e6f128a59423f4de3d6d4b5231ebe9aac2/src/validation.cpp#L1007
+pub struct BitcoinHalving;
+
+const HALVING_INTERVAL: u32 = 210_000;
+const INITIAL_ISSUANCE: u32 = 50;
+
+//TODO Do I need to account for decimal places here?
+impl Issuance<u32, u128> for BitcoinHalving {
+
+	fn issuance(block: u32) -> u128 {
+		let halvings = block / HALVING_INTERVAL;
+		// Force block reward to zero when right shift is undefined.
+		if halvings >= 64 {
+			return 0;
+		}
+
```

Comment:
Since we plan to halve issuance, starting from 50, should we use the permill or perbill denotation. That way we get more accuracy after the 3rd or 4th time its halved.

Reply:
I was wondering about this too. I don't think `Permill` is right because it represents numbers between 0 and 1, but I think you're right that we should handle this. The way both bitcoin and the balances pallet handle this is to just decide on a fixed number of digits that represent decimal places. In bitcoin it is 8 digits. We could make that change, but it would mean the issuance would be 5_000_000_000 untis per block initially. This would be much more convenient if we had a dedicated UI.